File: src/php/PhantasyIslandPhpRemoteControl/radio/TypeDef.php
<?php

namespace PhantasyIslandPhpRemoteControl\Radio;

class RadioCheckOptions
{
    public $frequencyMHz = null;
    public $txPowerDbm = null;
    public $rxSensitivityDbm = null;
    public $fresnelZoneRatio = null;
    public $skipFresnelZoneCheck = null;
    public $enableMultipath = null;
    public $maxReflectionPaths = null;
    public $maxReflectionPathLengthRatio = null;
    public $defaultTxAntennaGain_dBi = null;
    public $defaultRxAntennaGain_dBi = null;
    public $enableAntennaPattern = null;
    public $enableMutualCoupling = null;
    public $couplingNegligibleThresholdWavelengths = null;
    public $enableSINR = null;
    public $receiverBandwidthHz = null;
    public $minSINR_dB = null;
    public $enableFrequencyIsolation = null;
    public $enableNearFieldCorrection = null;
    public $enableNodeBodyOcclusion = null;

    public function to_dict()
    {
        $d = [];
        foreach (get_object_vars($this) as $k => $v) {
            if ($v !== null) $d[$k] = $v;
        }
        return $d;
    }
}

class RadioMaterialProperties
{
    public $id;
    public $displayName;
    public $penetrationLoss_dBPerMeter;
    public $reflectionCoefficient;
    public $defaultThickness_m;

    public function __construct($id, $displayName, $penetrationLoss, $reflection, $thickness)
    {
        $this->id = $id;
        $this->displayName = $displayName;
        $this->penetrationLoss_dBPerMeter = $penetrationLoss;
        $this->reflectionCoefficient = $reflection;
        $this->defaultThickness_m = $thickness;
    }

    public function to_dict()
    {
        return get_object_vars($this);
    }
}

class TypeDef
{
    public static function radio_check_options_from_dict($data)
    {
        $obj = new RadioCheckOptions();
        foreach ($data as $k => $v) {
            if (property_exists($obj, $k)) $obj->$k = $v;
        }
        return $obj;
    }

    public static function radio_material_properties_from_dict($data)
    {
        return new RadioMaterialProperties(
            $data['id'],
            $data['displayName'],
            $data['penetrationLoss_dBPerMeter'],
            $data['reflectionCoefficient'],
            $data['defaultThickness_m']
        );
    }
}

File: src/fs/TypeDef.fs
namespace PhantasyIslandRemoteControl.Radio

open System.Collections.Generic

type XYZ(x: double, y: double, z: double) =
    member val X = x with get, set
    member val Y = y with get, set
    member val Z = z with get, set

    member this.ToList() = [ this.X; this.Y; this.Z ]
    static member FromList(list: IList<double>) = XYZ(list.[0], list.[1], list.[2])

type RadioCheckOptions() =
    member val FrequencyMHz: double option = None with get, set
    member val TxPowerDbm: double option = None with get, set
    member val RxSensitivityDbm: double option = None with get, set
    member val FresnelZoneRatio: double option = None with get, set
    member val SkipFresnelZoneCheck: bool option = None with get, set
    member val EnableMultipath: bool option = None with get, set
    member val MaxReflectionPaths: double option = None with get, set
    member val MaxReflectionPathLengthRatio: double option = None with get, set
    member val DefaultTxAntennaGain_dBi: double option = None with get, set
    member val DefaultRxAntennaGain_dBi: double option = None with get, set
    member val EnableAntennaPattern: bool option = None with get, set
    member val EnableMutualCoupling: bool option = None with get, set
    member val CouplingNegligibleThresholdWavelengths: double option = None with get, set
    member val EnableSINR: bool option = None with get, set
    member val ReceiverBandwidthHz: double option = None with get, set
    member val MinSINR_dB: double option = None with get, set
    member val EnableFrequencyIsolation: bool option = None with get, set
    member val EnableNearFieldCorrection: bool option = None with get, set
    member val EnableNodeBodyOcclusion: bool option = None with get, set

    member this.ToDict() =
        let dict = Dictionary<string, obj>()
        let props = this.GetType().GetProperties()
        for prop in props do
            let value = prop.GetValue(this)
            if value <> null && prop.Name <> "ToDict" then
                let name = prop.Name.[0].ToString().ToLower() + prop.Name.Substring(1)
                match value with
                | :? XYZ as xyz -> dict.[name] <- xyz.ToList()
                | _ -> dict.[name] <- value
        dict

type RadioMaterialProperties() =
    member val Id = "" with get, set
    member val DisplayName = "" with get, set
    member val PenetrationLoss_dBPerMeter = 0.0 with get, set
    member val ReflectionCoefficient = 0.0 with get, set
    member val DefaultThickness_m = 0.0 with get, set

    member this.ToDict() =
        let dict = Dictionary<string, obj>()
        dict.["id"] <- this.Id
        dict.["displayName"] <- this.DisplayName
        dict.["penetrationLoss_dBPerMeter"] <- this.PenetrationLoss_dBPerMeter
        dict.["reflectionCoefficient"] <- this.ReflectionCoefficient
        dict.["defaultThickness_m"] <- this.DefaultThickness_m
        dict

type JoyStickInput() =
    member val Vx = 0.0 with get, set
    member val Vy = 0.0 with get, set
    member val Vz = 0.0 with get, set
    member val YawRate = 0.0 with get, set

    member this.ToDict() =
        let dict = Dictionary<string, obj>()
        dict.["vx"] <- this.Vx
        dict.["vy"] <- this.Vy
        dict.["vz"] <- this.Vz
        dict.["yawRate"] <- this.YawRate
        dict
